File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Power/DexterityPower.ts
import { SoundMaster } from "../Audio/SoundMaster";
import { PreModifyBlockMessage } from "../Events/PreModifyBlockMessage";
import FileHelper from "../FileHelper";
import { LocalizationProperty } from "../Gen/DB/Localization";
import { StringHelper } from "../StringHelper";
import { AbstractPower, PowerType } from "./AbstractPower";


export class DexterityPower extends AbstractPower{
    Name: string = LocalizationProperty.Read("能力-敏捷");
    Icon: string = FileHelper.FormatPath("Powers/strength.png");
    Id: string = "Strength";
    get Type(){return this.Amount >= 0 ? PowerType.Buff : PowerType.Debuff}
    GetDescription(...args: any[]): string {
        if(this.Amount >= 0){
            return StringHelper.FormatColorString(LocalizationProperty.Read("能力-敏捷(提升)")).format(this.Amount);
        }
        return StringHelper.FormatColorString(LocalizationProperty.Read("能力-敏捷(下降)")).format(this.Amount);
    }
    public PlayApplyPowerSfx() {
        SoundMaster.PlayVoice("POWER_DEXTERITY", 0.05);
    }
    PreModifyBlock(msg: PreModifyBlockMessage) {
        msg.block += this.Amount;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Power/Buff/DexterityPower.cs
using ModTheSpire.Event.Ids;
using TCG.Config;

namespace ModTheSpire.Power.Buff
{
    public class DexterityPower : AbstractPower
    {
        public override string Id { get; } = nameof(DexterityPower);
        public override string Name { get; } = LocalizationProperty.Read("能力-敏捷");
        public override string Icon { get; } = "Powers/strength.png";
        public override PowerType Type => Amount >= 0 ? PowerType.Buff : PowerType.Debuff;
        public override string GetDescription(object[] args)
        {
            return string.Format(LocalizationProperty.Read(this.Amount >= 0 ? "能力-敏捷(提升)" : "能力-敏捷(下降)"),this.Amount);
        }

        public override void PlayApplyPowerSfx()
        {
            base.PlayApplyPowerSfx();
            SoundMaster.PlayVoice(SoundMaster.POWER_DEXTERITY, 0.05f);
        }

        protected override void PreModifyBlock(PreModifyBlockMessage msg)
        {
            base.PreModifyBlock(msg);
            msg.Amount += Amount;
        }
    }
}